File: day10/joltage_adapters_bullers.py
from collections import Counter
import itertools

with open('day10/input') as f:
    input = [int(line) for line in f]

joltage_ratings = sorted([0] + input + [max(input) + 3])

difference_counts = Counter(
    [y - x for x, y in zip(joltage_ratings, joltage_ratings[1:])])

print(difference_counts[1] * difference_counts[3])

def tribonacci():
    a, b, c = 0, 1, 1
    while True:
        yield a
        a, b, c = (b, c, a + b + c)

def arrangements(ratings):
    run_len = 1
    result = 1
    for i, curr in enumerate(ratings):
        more = ratings[i + 1:]
        if not more:
            return result
        elif more[0] - curr == 1:
            run_len += 1
        else:
            nth_trib = next(itertools.islice(tribonacci(), run_len, None))
            result = result * nth_trib
            run_len = 1

print(arrangements(joltage_ratings))

File: day10/joltage_adapters_bullers.clj
(ns joltage-adapters
  (:require [clojure.string :as str]))

(def input
  (->> (str/split-lines (slurp "day10/input"))
       (map #(Integer/valueOf %))))

(def built-in-rating (+ (apply max input) 3))

(def joltage-ratings (sort (conj input 0 built-in-rating)))

(defn joltage-differences [joltage-ratings]
  (let [differences (map - (rest joltage-ratings) joltage-ratings)]
    (frequencies differences)))

(let [difference-counts (joltage-differences joltage-ratings)]
  (println (* (difference-counts 1) (difference-counts 3))))


(def tribonacci-seq
  ((fn trib [a b c]
     (lazy-seq (list* a (trib b c (+ a b c)))))
   0 1 1))

(defn arrangements [[curr & more] run-len result]
  (cond
    (nil? more) result
    (= (- (first more) curr) 1) (arrangements more (inc run-len) result)
    :else (arrangements more 1 (* result (nth tribonacci-seq run-len)))))

(println (arrangements joltage-ratings 1 1))
